src/cleaner/cleaner_main.py

```diff
@@ -49,12 +51,28 @@ def terminate(self):
                 self.logger.error("errors occur when terminating worker %s.", k)
                 self.logger.exception(e)
 
+    def update_liveness(self):
+        if self.liveness:
+            file_name = os.path.join("/tmp", self.liveness)
+            with open(file_name, "a"):
+                os.utime(file_name, None)
+
     def sync(self):
         try:
-            for w in self.workers.values():
-                w.join()
-        except:
+            while True:
+                stopped_workers = [(k, w) for k, w in self.workers.items() if not w.is_alive()]
+                if len(stopped_workers) > 0:
+                    for k, w in stopped_workers:
+                        self.logger.error("worker %s exit with code %s", k, w.exitcode)
+                        self.workers.pop(k)
+                if len(self.workers) == 0:
+                    self.logger.info("all workers are stopped and exit cleaner.")
+                    break
+                self.update_liveness()
+                time.sleep(2)
+        except Exception as e:
             self.logger.error("cleaner interrupted and will exit.")
+            self.logger.exception(e)
```

Comment:
[logger.exception](https://docs.python.org/2/library/logging.html#logging.Logger.exception) method accept string msg, and will print msg and exception.
use like `self.logger.exception(cleaner interrupted and will exit.)`

Reply:
change done